### PR TITLE
Use the Parsoid v3 API

### PR DESCRIPTION
**Pull request by cscott**
Needs love & attention wrt to the sections API.  Parsoid v3 returns section offsets without the <body> wrapper; some of this code might need to be updated accordingly?

### CODE REVIEW

**Comment by cscott** [comment]
Parsoid v3 has been deployed now.

File: mods/parsoid.js [hunk 1]
```diff
@@ -618,7 +614,7 @@ PSP.makeTransform = function(from, to) {
             var innerRes = res.body[to];
             innerRes.status = 200;
             // Handle bodyOnly flag
-            if (to === 'html' && req.body.bodyOnly) {
+            if (from === 'sections' && to === 'html' && req.body.bodyOnly) {
```
Comment (Pchelolo) [inline mods/parsoid.js]: Why is that added? In the [docs](https://rest.wikimedia.org/en.wikipedia.org/v1/?doc#!/Transforms/post_transform_html_to_html_title_revision) we state that 'bodyOnly' is valid for other `from` formats as well.

File: mods/parsoid.js [hunk 2]
```diff
@@ -618,7 +614,7 @@ PSP.makeTransform = function(from, to) {
             var innerRes = res.body[to];
```
Comment (Pchelolo) [inline mods/parsoid.js]: The bug is around here: for `to === 'wikitext'` Parsoid returns an unwrapped response.  From what I can tell from testing, this should look like

```
var innerRes = {};
if (to === 'wikitext') {
    innerRes.body = res.body;
    innerRes.headers = {
        'content-type': res.headers['content-type']
    }
} else {
    innerRes = res.body[to];
}
```
And also the content-type header is different from what we expect, the charset is the second param, not the third.

**Comment by Pchelolo** [comment]
@cscott Unfortunately the tests still fail. I've left a couple of comments inline

File: mods/parsoid.js (hunk 1 above)
Comment (cscott) [inline mods/parsoid.js]: Because you pass `bodyOnly` along in the request to Parsoid, and (in v3) it has already stripped the body tag from the output when you get it back.

But the `sections` stuff generates its own body tag around the output, which then has to be manually stripped here.  This is a bug, but like I said, I don't fully understand what the `sections` stuff is doing and could use some help.  I'm worried I could screw up the offsets if I just blindly remove `<body>` from the code which generates the sections output.

File: mods/parsoid.js (hunk 2 above)
Comment (cscott) [inline mods/parsoid.js]: Yes, express reorders the charset.  Can't be helped.

**Comment by cscott** [comment]
Updated the patch and the tests pass.  Could still use some review of the `sections` stuff, since I suspect that the need for the `cheapBodyInnerHTML` function could be removed entirely with some refactoring.  But perhaps that can be a follow-up patch by someone who understands `sections` better than I.

File: mods/parsoid.js [hunk 3]
```diff
@@ -615,10 +611,18 @@ PSP.makeTransform = function(from, to) {
         return transform
         .then(function(res) {
             // Unwrap to the flat response format
-            var innerRes = res.body[to];
+            var innerRes = {};
+            if (to === 'wikitext') {
```
Comment (gwicke) [inline mods/parsoid.js]: Why is there a special case for wikitext?

File: mods/parsoid.js [hunk 4]
```diff
@@ -615,10 +611,18 @@ PSP.makeTransform = function(from, to) {
...
+                innerRes.headers = {
+                    'content-type': res.headers['content-type']
+                };
+            } else {
+                innerRes = res.body[to];
+            }
             innerRes.status = 200;
             // Handle bodyOnly flag
-            if (to === 'html' && req.body.bodyOnly) {
+            if (from === 'sections' && to === 'html' && req.body.bodyOnly) {
```
Comment (gwicke) [inline mods/parsoid.js]: Why is this limited to `from === 'sections'`?

Comment (cscott) [inline mods/parsoid.js]: See above: because the section extraction code adds a <body> around its result, which then has to be stripped off.  Parsoid no longer does this.  Help is requested to fix up the section extraction code, since I don't understand it well enough (and my daughter is going to be born within a matter of hours now).

File: mods/parsoid.js (hunk 3 above)
Comment (cscott) [inline mods/parsoid.js]: It should probably be `if (to !== 'pagebundle')` instead.  The `html` and `wikitext` formats return the results in the main body of the response (as does RESTBase).  It's only the `pagebundle` format that returns a JSON object wrapping the content instead.  Again, help refactoring would be appreciated if you want to turn the default case around and specially-handle only the pagebundle case.

File: config.test.yaml [hunk 5]
```diff
@@ -265,7 +265,7 @@ test:
   content_types:
     html: text/html;profile="mediawiki.org/specs/html/1.1.0";charset=utf-8
     data-parsoid: application/json;profile="mediawiki.org/specs/data-parsoid/0.0.1"
-    wikitext: text/plain;profile="mediawiki.org/specs/wikitext/1.0.0";charset=utf-8
+    wikitext: text/plain; charset=utf-8; profile="mediawiki.org/specs/wikitext/1.0.0"
```
Comment (gwicke) [inline config.test.yaml]: This seems like a random change. Could you keep the content-type as emitted by Parsoid the same?

Comment (cscott) [inline config.test.yaml]: No, it's not me reordering it, it's express.  It would require an upstream patch to express.

Comment (gwicke) [inline config.test.yaml]: I see. In that case, I guess we should change our content types across the board, and create similar normalization logic for compatibility checks.

Comment (cscott) [inline config.test.yaml]: You can use the same `content-type` npm module to do the normalization (and the comparison, I think).

Comment (gwicke) [inline config.test.yaml]: This is now tracked in https://phabricator.wikimedia.org/T116333.

**Comment by Pchelolo** [comment]
This PR is outdated, closing it in favour of https://github.com/wikimedia/restbase/pull/395
